torch/_decomp/decompositions.py

```diff
@@ -1782,6 +1782,7 @@ def nop_decomposition(x):
 # Also register to the Autograd dispatch key, so this decomp can run above autograd.
 # native_batch_norm needs to decompose into other ops before autograd.
 @aten.cudnn_batch_norm.default.py_impl(DispatchKey.Autograd)
+@aten.cudnn_batch_norm.default.py_impl(DispatchKey.CompositeImplicitAutograd)
```

Comment:
Wait @bdhirsh is this a safe thing to do? Now even in eager mode, you will run through this implementation, right? cudnn_bn has it own fw and bw impl https://fburl.com/code/8t71v9nj

Reply:
@kimishpatel so what's going on here is that we are registering this decomposition to the **python dispatcher** (using `.py_impl`). The python dispatcher is opt-in, and can be turned on using `with enable_python_dispatcher()`.

In eager mode we do not turn it on, but during compilation in PT2 we always turn it on (this has been very convenient for changes decomps at the autograd level that have been painful for torch.compile, without having to worry about BC-breaking eager mode).

> is this a safe thing to do

Based on everything I said, a very reasonable q is "ok, but will this change the result of batch norm in eager vs. compile?".

We're relying here on the fact that we expect `torch.batch_norm` to give (nearly) the same output, regardless of whether or not cudnn was used. cudnn won't give bitwise equivalent outputs compared to our native batchnorm kernel, but compile already doesn't promise to be bitwise equivalent to eager mode - so we're basically saying here that "even if we would have run cudnn if the user called batch norm in eager mode, we will skip cudnn and use inductor if you are using compile".

Reply:
Hmm. I don't know what inductor issue you ran into, but one thing to mention is that if you want the decomp to run **before** autograd runs, then the decomp needs to be autograd-preserving.


For export though, it probably doesn't matter if this decomp runs before/after autograd, if you only care about inference.

Reply:
@larryliu0820 what was the inductor issue you ran into?

Reply:
Here: https://github.com/pytorch/pytorch/actions/runs/6018775992/job/16327865852

The PR commit is https://github.com/pytorch/pytorch/pull/108180/commits/e9818631222a7b595b4fbddf43cb80c6bf7687a4

Reply:
Oh that error is fixable - you probably just need to talk to some folks from inductor (maybe Elias). Inductor is complaining because there is now a fallback *and* a decomposition defined for that flash attention op, so inductor does'nt know which one to use.